pkg/controller/admissionchecks/preemption/preemption.go

```diff
@@ -0,0 +1,301 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package preemption
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/constants"
+	"sigs.k8s.io/kueue/pkg/scheduler/preemption"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+const (
+	controllerName  = "PreemptionController"
+	throttleTimeout = 500 * time.Millisecond
+)
+
+type updateStatusFnc func(context.Context, client.Client, *kueue.Workload, bool) error
+
+// Controller manages the preemption process of the workloads after quota reservation taking into account
+// its admission checks states. It will issue eviction for other workloads, and mark the preemption as
+// successful or not based on the current state of the resources pool. Check run() for more details.
+// The controller will react to workload and admission check changes, check NotifyWorkloadUpdate()
+// and NotifyAdmissionCheckUpdate() for details.
+type Controller struct {
+	log         logr.Logger
+	cache       *cache.Cache
+	client      client.Client
+	recorder    record.EventRecorder
+	preemptor   *preemption.Preemptor
+	ctx         context.Context
```

Comment:
Can we avoid to have `ctx` in this `Controller`?

https://pkg.go.dev/context#pkg-overview

> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx: